src/psm/src/connection.cpp

```diff
@@ -0,0 +1,262 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "connection.h"
+
+#include "node.h"
+#include "odb/db.h"
+
+namespace psm {
+
+Connection::Connection(Node* node0, Node* node1) : node0_(node0), node1_(node1)
+{
+  ensureNodeOrder();
+}
+
+void Connection::ensureNodeOrder()
+{
+  if (node0_ == nullptr) {
+    std::swap(node0_, node1_);
+
+    return;
+  }
+
+  if (node1_ == nullptr) {
+    std::swap(node0_, node1_);
+
+    return;
+  }
+
+  if (!node0_->compare(node1_)) {
+    std::swap(node0_, node1_);
+  }
+}
+
+Node* Connection::getOtherNode(const Node* node) const
+{
+  if (node == node0_) {
+    return node1_;
+  }
+  return node0_;
+}
+
+Connection::Conductance Connection::getConductance(
+    const ResistanceMap& res_map) const
+{
+  const auto res = getResistance(res_map);
+  if (res == 0) {
+    return std::numeric_limits<Conductance>::infinity();
+  }
+  return 1.0 / res;
+}
+
+void Connection::changeNode(Node* orgnode, Node* newnode)
+{
+  if (node0_ == orgnode) {
+    node0_ = newnode;
+  }
+
+  if (node1_ == orgnode) {
+    node1_ = newnode;
+  }
+}
+
+void Connection::print(utl::Logger* logger) const
+{
+  logger->report(describe());
+}
+
+int Connection::getDBUs() const
+{
+  const Node* node = node0_ ? node0_ : node1_;
+  if (node) {
+    return node->getLayer()->getTech()->getDbUnitsPerMicron();
+  }
+  return 1;
+}
+
+bool Connection::compare(const Connection* other) const
+{
+  return compareTuple() < other->compareTuple();
+}
+
+bool Connection::compare(const std::unique_ptr<Connection>& other) const
+{
+  return compare(other.get());
+}
+
+std::tuple<int, int, int, int, int, int> Connection::compareTuple() const
+{
+  int node0_x = 0;
+  int node0_y = 0;
+  int node0_l = 0;
+
+  int node1_x = 0;
+  int node1_y = 0;
+  int node1_l = 0;
+
+  if (node0_ != nullptr) {
+    const auto& pt = node0_->getPoint();
+    node0_x = pt.getX();
+    node0_y = pt.getY();
+    node0_l = node0_->getLayer()->getNumber();
+  }
+
+  if (node1_ != nullptr) {
+    const auto& pt = node1_->getPoint();
+    node1_x = pt.getX();
+    node1_y = pt.getY();
+    node1_l = node1_->getLayer()->getNumber();
+  }
+
+  return {node0_l, node0_x, node0_y, node1_l, node1_x, node1_y};
```

Comment:
Would (a,b) != (b,a) be correct?

Reply:
the `ensureNodeOrder` function should force the order of the nodes, so I don't have to account for `b,a` case.